# Mathematical Modeling Homework: Chapter 8: 4, 5, 6

### <center> Mitch Lowry <br> April 17, 2016 </center>

**Problem 4:**  Reconsider the inventory problem of Example 8.1, but now suppose that three additional aquariums are ordered any time that there are less than two in stock at the end of the week.

<br>

<center> Background information from Example 8.1 </center>

<br>

A pet store sells a limited number of 20-gallon aquariums. At the end of each week, the store manager takes inventory and places orders. Store policy is to order three new 20-gallon aquariums at the end of the week if all the current inventory has been sold. If even one of the 20-gallon aquariums remains in stock, no new units are ordered. This policy is based on the observation that the store only sells an average of one of the 20-gallon aquariums per week. Is this policy adequate to guard against potential lost sales of 20-gallon aquariums due to a customer requesting one when they are out of stock?

__Part (a): Determine the probability that demand exceed supply on any given week. Use the five-step method and model as a Markov chain in steady state.__

### Step 1: Ask a question

Our question here is what is the probability that demand exceeds supply. In order to do this we must make some assumptions about the arrival of buyers who will purchase an aquarium if there is one in stock and the supply of aquariums at the beginning of any given week, which we define as $D_n$ and $S_n$, respectively. Our assumptions about the supply of aquariums have already been made by the store's current inventory policy, which we may or may not alter to improve profits. We assume that the probability distribution of the demand for aquariums is a poisson distribution centered at 1. A summary of step one is provided below.

<br>

**Variables:**

$S_n =$ supply of aquariums at the beginning of week $n$.

$D_n =$ demand for aquariums during week $n$.


**Assumptions:**

If $D_{n-1} < S_{n-1} - 1$, then $S_n = S_{n-1} - D_{n-1}$.

If $D_{n-1} \ge S_{n-1} - 1$, then $S_n = S_{n-1} - D_{n-1} + 3$.

$Pr(D_n = k) = \dfrac{e^{-1}}{k!}$


**Objective:** 

Calculate $Pr(D_n > S_n)$.

<br>

### Step 2: Select a Modeling Approach

As designated by the problem, we will model this system as a Markov chain. 

### Step 3: Formulate the Model

Here we are considering the random variable $S_n$. We know that $S_n \in \{2, 3, 4\}$ because at the end of week $n-1$ there could potentially be $0$ or $1$ aquariums left ($S_{n-1} - D_{n-1} \in \{0, 1\}$), which would mean $S_n \in \{3, 4\}$, and otherwise we have $2$ or more aquariums left at the end of week; however, it it not possible to have more than $4$ aquariums in stock. Since we have assumed that $D_n$ is a poisson process centered at $1$, $S_n$ depends only on $S_{n-1}$. Thus, we can say that the sequence $\{S_n\}$ is a Markov chain.

We do not know the value of $S_0$, but it seems reasonable to make the assumption that $S_0 = 3$; however, we may adjust this assumption later on. 

The only way to get to state $j$ at time $n$ is to have come from some state $i$ at time $n-1$. We assume that the probability of changing between any two states is constant, i.e. it does not depend on time. We define $p_{ij}$ as the probability of going from state $i$ to state $j$.

First of all, the probability of transitioning between states is clearly dependent on the distribution of $D_n$. We compute relevant values of $D_n$ below.

In [1]:
import math

def D(r, k):
    return r**k*math.e**(-r)/math.factorial(k)

# We will define m as the largest value of potential buyers that we are interested in.
# In this case we have m = 4.

vals = []
r = 1
m = 4
k = 0
while k <= m:
    vals.append(D(r, k))
    k += 1

vals.append(1-sum(vals))
    
for num in range(5):
    print(u'P(D\u2099 = ' + str(num) + ") = " + str(vals[num]) + ".")
    
    
print(u"P(D\u2099 > " + str(m) + ")" +  " = "  + str(vals[len(vals) - 1]) + ".")
print("As a check, the sum of these values is " + str(sum(vals)) + ".")

P(Dₙ = 0) = 0.36787944117144233.
P(Dₙ = 1) = 0.36787944117144233.
P(Dₙ = 2) = 0.18393972058572117.
P(Dₙ = 3) = 0.06131324019524039.
P(Dₙ = 4) = 0.015328310048810098.
P(Dₙ > 4) = 0.00365984682734366.
As a check, the sum of these values is 1.0.


We can now use these probabilites to find the transition probabilites.

We can see the following:

* $p_{22} = P(D_n = 0)$.
* $p_{23} = P(D_n \ge 2) = 1 - P(D_n < 2) = 1 - P(D_n = 0) - P(D_n = 1)$.
* $p_{24} = P(D_n = 1)$.
* $p_{32} = P(D_n = 1)$.
* $p_{33} = P(D_n \ne 1) = 1 - P(D_n = 1)$.
* $p_{34} = 0$.
* $p_{42} = P(D_n = 2)$.
* $p_{43} = P(D_n = 1) + P(D_n \ge 4)$.
* $p_{44} = P(D_n = 0) + P(D_n = 3)$.

We define our probability transition matrix as P. We will compute each of these in order below.

In [2]:
P = []
P.append([vals[0], 1 - vals[0] - vals[1], vals[1]])
P.append([vals[1], 1 - vals[1], 0])
P.append([vals[2], vals[1] + vals[4] + vals[5], vals[0] + vals[3]])
print("P = " + str(P) + ".")

P = [[0.36787944117144233, 0.26424111765711533, 0.36787944117144233], [0.36787944117144233, 0.6321205588285577, 0], [0.18393972058572117, 0.3868675980475961, 0.4291926813666827]].


### Step 4: Solving the Model

It is clear that $\{S_n\}$ is ergodic; thus, we know that there is a unique steady-state probability vector $a$ that can be computed by solving the steady-state equations. Let $a_2$ be the probability of being in state $2$ in the long-run, $a_3$ be the probability of being in state $3$ in the long-run, and $a_4$ be the probability of being in state $4$ in the long-run. We know that $a = a \cdot P$, and also that $a_2 + a_3 + a_4 = 1$. We can see that we have $4$ equations and $3$ variables here, so we can remove one of the equations. We solve this using the SymPy library below.

In [3]:
from sympy import *

a = []
a2, a3, a4 = symbols('a2 a3 a4')
set_a = solve((a2 + a3 + a4 - 1, a2 - P[0][0]*a2 - P[1][0]*a3 - P[2][0]*a4, \
          a3 - P[0][1]*a2 - P[1][1]*a3 - P[2][1]*a4), (a2, a3, a4))

a.extend([float(set_a[a2]),float(set_a[a3]), float(set_a[a4])])

print(str(set_a))
print(str(a))

# For some reason sympy is being weird and storing the variables out of order.
# Let's check our answers to be sure that they are correct. Due to rounding we will
# use some small delta value instead of using 0.

if a[0] + a[1] + a[2] - 1 < .001 and a[0] - P[0][0]*a[0] - P[1][0]*a[1] - \
P[2][0]*a[2] < .001 and a[1] - P[0][1]*a[0] - P[1][1]*a[1] - P[2][1]*a[2] < .001:
    print("I guess it worked. It just solved them out of order.")
else:
    print("The solve function is finicky.")

{a2: 0.328890387984264, a4: 0.211966469575059, a3: 0.459143142440677}
[0.3288903879842639, 0.459143142440677, 0.21196646957505913]
I guess it worked. It just solved them out of order.


The main objective here, as outlined in step one is to compute $P(D_n > S_n)$. By probability rules we know that $$P(D_n > S_n) = \sum_{i=2}^{4} P(D_n > S_n | S_n = i) \cdot  P(S_n = i) = \sum_{i=2}^{4} P(D_n > S_n | S_n = i) \cdot  a_i$$.

We compute $P(D_n > S_n | S_n =i)$ below.

In [4]:
P_excess_demand_state_i = []

for i in range(2, 5):
    P_excess_demand_state_i.append(1 - sum(vals[:i+1]))
    #print(str(vals[:i+1])) Checking stuff
                        
P_excess_demand_state_i

[0.08030139707139416, 0.01898815687615374, 0.00365984682734366]

Now we compute the $P(D_n > S_n)$ by taking the dot product of $a$ and the above vector.

In [5]:
print(u"We have P(D\u2099 > S\u2099) = " + \
      str(sum(map(lambda x,y:x*y,P_excess_demand_state_i,a))))

We have P(Dₙ > Sₙ) = 0.03590440446694056


### Step 5: Sensitivity Analysis

<br>

For our sensitivity analysis we will not talk about seasonal trends or anything requiring the Markov model to depend on more than $S_{n-1}$. Like in Example $8.1$ we will examine the effect of varying the average number of potential buyers per week ($\lambda$). To do this, I was kind of lazy. I simply reran the above code boxes, except I changed the definition of the probability distribution of $D_n$. That is, I changed the -1 to -.8, -.9, -1.1, and -1.2, and simply wrote down each of the values. The point with choosing numbers that are pretty close to 1 is that the store will probably not have huge changes in the average number of customers per week over time; however, it could be the case that the original problem had an unreasonable number for the number of potential customers per week.

I ended up that the probability of excess demand as .0202, .0275, .0359, .0455, .0563 for when the probability distribution of $D_n$ is centered at the values .8, .9, 1.0, 1.1, and 1.2, respectively. More sophisticated measurements of sensitivity can be derived from these numbers, if desired.

Note that the probability of excess demand becomes very large $\lambda$ values that might seem unreasonable. For example, if the store doubles it's potential customers ($\lambda = 2$), then the probability of excess demand is .1774. Having $\lambda$ double may seem unreasonable, but if there are seasonal trends, then this may very well be the case. That is, $\lambda=1$ were based off of the average for the whole year, and the number of aquariums sold weekly in the summer is double that of the whole year, then during the summer we would have a large probability of excess demand.

__Part (b): Use the steady-state probabilities from part (a) to calculate the expected number of aquariums sold per week under this inventory policy.__

<br>

Let the random variable $X$ represent the number of aquariums sold per week. Then we have the following:

<br>

$E(X) = \sum_{i=0}^{4} i \cdot P(X = i) = \sum_{i=0}^{4} i \cdot P($There are at least i tvs in stock on any given week $ \cap $ The demand for tvs is i on any given week$) = \sum_{i=0}^{4} i \cdot P($There are at least i tvs in stock on any given week$) \cdot P($The demand for tvs is i on any given week$) = \sum_{i=0}^{4} i \cdot P(S_n \ge i \cap D_n = i) = \sum_{i=0}^{4} i \cdot P(S_n \ge i) \cdot$  $P(D_n = i) =$ $0 \cdot (a_2 + a_3 + a_4) \cdot P(D_n = 0)$ $+$ $1 \cdot (a_2 + a_3 + a_4) \cdot P(D_n = 1)$ $+$ $2 \cdot (a_2 + a_3 + a_4) \cdot P(D_n = 2)$ $+$ $3 \cdot (a_3 + a_4) \cdot P(D_n = 3)$ $+$ $4 \cdot (a_4) \cdot P(D_n = 4)$.

A step of explanation is missing here, but it is clear that they can sell at most as many aquariums as they have in stock, and we determine the number of aquariums in stock by the long-term probabilities of being in a state at any given time. This computation is shown below.

In [6]:
EX = 0
i = 0
start = 0
while i < 5:
    if i > start + 2:
        start += 1
    EX += sum(a[start:])*vals[i] * i
    i += 1
    
print("The expected number of aquariums sold per week is " + str(EX) + ".")

The expected number of aquariums sold per week is 0.8721989479218436.


__Part (c): Repeat part (b) for the inventory policy in Example 8.1__

<br>

In this case, we repeat the process done in the previous code box, except we have a different steady-state probability vector. This vector, call it $v = (v_1, v_2, v_3) = (0.285,  0.263,  0.452)$. This calculation is shown below.

In [7]:
v = [.285, .263, .452]
EX2 = 0
i = 0
start = 0
while i < 5:
    if i > start + 1:
        start += 1
    EX2 += sum(v[start:])*vals[i] * i
    i += 1
    
print("The expected number of aquariums sold per week is " + str(EX2) + ".")

The expected number of aquariums sold per week is 0.7140539953137696.


__Part (d): Suppose that the store makes a profit of 5 dollars per 20-gallon aquarium sold. How much would the store gain by implementing the new inventory policy?__

<br>

We will compute the weekly and 52-weekly (approximately yearly) gain in profit below.

In [8]:
weekly_gain = 5*(EX - EX2)
yearly_gain = 52*weekly_gain
print("The gain in weekly profit is " + str(weekly_gain) + " dollars.")
print("The gain in yearly profit is " + str(yearly_gain) + " dollars.")

The gain in weekly profit is 0.7907247630403702 dollars.
The gain in yearly profit is 41.11768767809925 dollars.


**Problem 5:**  For the purposes of this problem, we will consider the stock market to be in one of three states:

<br>

<center>
<ol>
    <li> Bear market </li>
    <li> Strong bull market </li>
    <li> Weak bull market </li>
</ol>
</center>

<br>

Historically, a certain mutual fund gained -3%, 28%, and 10% annually when the market was in states 1, 2, and 3 respectively. Assume that the state transition probability matrix

<br>

<center>
$P = \begin{bmatrix}
    p_{11} & p_{12} & p_{13} \\
    p_{21} & p_{22} & p_{23} \\
    p_{31} & p_{32} & p_{33}
\end{bmatrix} 
=
\begin{bmatrix}
    0.90 & 0.02 & 0.08 \\
    0.05 & 0.85 & 0.10 \\
    0.05 & 0.05 & 0.90
\end{bmatrix}$
</center>

<br>

applies to the weekly change of state in the stock market.

**Part (a): Determine the steady-state distribution of the market state.**

<br>

Let $b_i$ be the probability of being in state $i$ in the long-run. Let $b = [b_1, b_2, b_3]$. As before, this is an ergodic Markov chain, so we have $a = aP$. We solve this system below.

In [16]:
from sympy import *

P = [[0.90, 0.02, 0.08],[0.05, 0.85, 0.10],[0.05, 0.05, 0.90]]
b = []
b1, b2, b3 = symbols('b1 b2 b3')
set_b = solve((b1 + b2 + b3 - 1, b1 - P[0][0]*b1 - P[1][0]*b2 - \
               P[2][0]*b3, b2 - P[0][1]*b1 - P[1][1]*b2 - \
               P[2][1]*b3), (b1, b2, b3))

#Let's fix these orderings again.

b.extend([float(set_b[b1]), float(set_b[b2]), float(set_b[b3])])
print(set_b)
print(b)

{b2: 0.200000000000000, b1: 0.333333333333333, b3: 0.466666666666667}
[0.3333333333333333, 0.2, 0.4666666666666667]


We can see that $b = [b_1, b_2, b_3] = [\dfrac{1}{3}, \dfrac{2}{10}, \dfrac{7}{15}]$

**Part (b) Suppose that 10,000 dollars is invested in the fund for ten years. Determine the expected total yield. Does the order of state transition make any difference?**

<br>

Here we have a compounding interest problem. We let the random variable $Y_n$ be the amount of money that is in the fund at year $n$, and we know that $Y_0 = 10000$. Let the random variable $r_n$ be the interest rate recieved in year $n$. Additionally, since we have compounding interest we have $Y_n = Y_{n-1} + r_{n-1} \cdot Y_{n-1}$. 

**Part (c): In the worst-case scenario, the long-run expected proportion of time in each market state is 40, 20, and 40 percent, respectively. What is the effect on the answer to part (b)?**

<br>

**Part (d): (d) In the best-case scenario, the long-run expected proportion of time in each market state is 10, 70, and 20 percent, respectively. What is the effect on the answer to part (b)?**

<br>

**Part (e) Does this mutal fund offer a better investment opportunity than a money market fund currently yielding about 8 percent? Consider that the money market fund offers a lower risk.**

<br>

**Problem 6:** A Markov chain model of floods uses the state variable $X_n = 1, 2, 3, 4, 5$ where state $1$ means the average daily flow is below 1,000 cubic feet per second (cfs), state $2$ is 1,000-2,000 cfs, state $3$ is 2,000-5,000 cfs, $4$ is 5,000-10,000 cfs, and $5$ is over 10,000 cfs. The state transition probability matrix for this model is

<br>

<center>
$P = \begin{bmatrix}
    p_{11} & p_{12} & p_{13} & p_{14} & p_{15} \\
    p_{21} & p_{22} & p_{23} & p_{24} & p_{25} \\
    p_{31} & p_{32} & p_{33} & p_{34} & p_{35} \\
    p_{41} & p_{42} & p_{43} & p_{44} & p_{45} \\
    p_{51} & p_{52} & p_{53} & p_{54} & p_{55}
\end{bmatrix} 
=
\begin{bmatrix}
    0.90 & 0.05 & 0.025 & 0.015 & 0.01 \\
    0.30 & 0.70 & 0 & 0 & 0 \\
    0 & 0.40 & 0.60 & 0 & 0 \\
    0 & 0 & 0.60 & 0.40 & 0 \\
    0 & 0 & 0 & 0.80 & 0.20
\end{bmatrix}$.
</center>

__Part (a): Draw the state transition probability diagram for this model.__

I have attached this on a piece of copy paper at the end of this document.

__Part (b): Find the steady-state probability distribution for this model.__

Again, this is a ergodic Markov chain. We know that they steady-state probability column vector $b$ is such that $b = bP$. We will recycle the code used at the top of the document to do this.

In [9]:
from sympy import *

P = [[0.90, 0.05, 0.025, 0.015, 0.01],[0.30, 0.70, 0, 0, 0],\
     [0, 0.40, 0.60, 0, 0], [0, 0, 0.60, 0.40, 0], [0, 0, 0, 0.80, 0.20]]
b = []
b1, b2, b3, b4, b5 = symbols('b1 b2 b3 b4 b5')
set_b = solve((b1 + b2 + b3 + b4 + b5 - 1, b1 - P[0][0]*b1 - P[1][0]*b2 - P[2][0]*b3 - \
               P[3][0]*b4 - P[4][0]*b5, b2 - P[0][1]*b1 - P[1][1]*b2 - P[2][1]*b3 - \
              P[3][1]*b4 - P[4][1]*b5, b3 - P[0][2]*b1 - P[1][2]*b2 - P[2][2]*b3 - \
              P[3][2]*b4 - P[4][2]*b5, b4 - P[0][3]*b1 - P[1][3]*b2 - P[2][3]*b3 - \
              P[3][3]*b4 - P[4][3]*b5), (b1, b2, b3, b4, b5))

#Again, the answers are out of order, so we'll fix this.

b.extend([float(set_b[b1]), float(set_b[b2]), float(set_b[b3]), float(set_b[b4]), \
          float(set_b[b5])])

print(type(set_b[b1]))
print(str(set_b))
print("As a check, we can see that the sum off the steady-state probabilities is " + \
     str(sum(b)) + ".")
print(str(b))

<class 'sympy.core.numbers.Float'>
{b5: 0.00826446280991736, b3: 0.0826446280991736, b4: 0.0275482093663912, b1: 0.661157024793388, b2: 0.220385674931129}
As a check, we can see that the sum off the steady-state probabilities is 1.0.
[0.6611570247933884, 0.22038567493112948, 0.08264462809917356, 0.027548209366391185, 0.008264462809917356]


Thus we have that $b =[b_1, b_2, b_3, b_4, b_5] = [0.6612, 0.2204, 0.0826, 0.0275, 0.0083
]$.

**Part (c): How often are severe flood (over 10,000 cfs) expected to occur?**

We can see from the steady-state probabilites that in the long-run the river is in a state that would be classified as a severe flood $0.83\%$ of the time.

** Part (d): A reservoir used for drought storage depends on the flow of this river. When the flow is over 5,000 cfs, the reservoir is allowed to store 1,000 acre-feet per day. When the flow is under 1,000 cfs, the reservoir is required to release 100 acre-feet per day back into the river. Find the expected annual number of acre-feet of water stored in the reservoir. Is this number positive or negative? What does this mean?**

Let $R_n$ be a random variable representing the amount of water stored in the reservoir on day $n$. It is clear that $E(R_n) = 1000 \cdot P(X_n \ge 4) - 100 \cdot P(X_n < 4) = 1000 \cdot (b_4 + b_5) - 100 \cdot b_1$. To find the expected value for the year, we multiply this number by $365.25$, as shown below.

In [10]:
expected_daily_water = 1000*(b[3]+b[4]) - 100*(b[0])
expected_annual_water = 365.25*expected_daily_water

print(expected_daily_water)
print("The expected annual number of acre-feet of water stored in the reservoir is " + \
     str(expected_annual_water) + ".")

-30.30303030303031
The expected annual number of acre-feet of water stored in the reservoir is -11068.181818181822.


This is an extremely large negative number, and what it means is that the river is low very often, i.e. the probability of the flow of the river being less than 1000 cfs is so high that the reservoir frequently runs out of water, which means there are large periods of time during which the reservoir is not usable, and drought may insue. However, having a reservoir is better than not having one at all, I suppose. Also, it may just be the case that the river frequently has less than $1000$ cfs, but that there is not necessarily a drought, meaning that most of the time, the reservoir is not needed, and that it functions as a way to reduce floods and reintroduce the water at a safe but also potentially beneficial point in time.